### ISSUE TITLE
Legacy data interpreter

### ISSUE THREAD

**Issue by accursedCursive** [endless-sky/endless-sky#3609]
One thing getting in the way of ES development is that any changes (not additions) to the tag system would require countless data files to be changed.
It would be useful if data files didn't have to be changed, and instead were reinterpreted as the game loaded. There'd be no worry about breaking compatibility with anything as you deprecate tags.
It means that you don't have to put as much work to test out alterations to the game's code, and plugins are safe.

For example, say that a new "hardpoint" class was added.
It would be used like this:
```
hardpoint "Gunner Window"
	cone 90
	"never disabled"
	priority 2 # only active when the crew isn't too busy with priority 1 hardpoints
	"install on purchase" 0 # meaning that instead of installing on purchase, when the ship launches compatible outfits that are not already installed in a hardpoint
	capacity 3 # can have 3 regular outfits per hardpoint, meaning you don't have to have 20 gunner windows for a huge ship which would clutter data files
	opportunistic

outfit "Laser Rifle"
	hardpoint "Gunner Window" 1
	"required crew" 1

outfit "Gatling Gun"
	hardpoint "Gun Port" 1 # prefers gun ports but can be installed in a gunner window instead
	"required crew" 0 # the first crew requirement after the chosen hardpoint is used
	hardpoint "Gunner Window" 3 # the 3 means it takes up the whole gunner window
	"required crew" 2

ship "Modified Argosy"
hardpoint "Gunner Window" -40 20 270
hardpoint "Gunner Window" 40 20 90
```

Obviously, this would mean that the game now expects to see:
hardpoint "Gun Port" 1
instead of:
"gun ports" -1

So, the job of the legacy data interpreter would be to interpret every occurrence of "gun ports" -1 as if it were the new hardpoint system.
There's plenty of other ways this could be used, and it means that the first way a feature has to be implemented doesn't have to be the final way it's implemented.
Being able to modify stuff after it's released is vitally important for improving code.

**Comment by tehhowch** [comment]
A somewhat-similar idea was offered in the past (introducing a version number in a save file so the appropriate loading constructs can be used, #2831), and the answer is essentially "let's cross that bridge when we come to it." I am in agreement with this concept of not implementing a legacy code loader when there is no legacy code to be loaded.

An Issue that describes a useful and well-thought-out feature (or a pull request implementing such a thing) that requires a breaking change in load behavior is not impossible to merge. However (and I feel rightly so), it will not merge quickly: alternatives that could achieve the same result will be sought.
(In my opinion, for a breaking change proposal to be successful, the person who is doing the proposing is going to be the one who needs to do the bulk of the work - it's one thing to ask for a change that would allow one to do X, and its quite another to have already prepared X and only need to agree on how the information goes from hard drive files to game objects.

**Comment by accursedCursive** [comment]
The reason there's no legacy code to be loaded is because no one will make anything legacy code, because there's no system in place to handle legacy code.
Someone has to make the first move.

As you say, a really good feature could be that first move, but the things that would break compatibility would be the things that allow really good features to be implemented without being impressive themselves.
For example, a hardpoint class. It's a boring feature by itself, and doesn't add anything new to the game. What it does do is it allows modders to add new things to the game, which in turn can enter the vanilla game, but people don't plan far ahead enough to care if it doesn't immediately affect the vanilla game.
Maybe what someone needs to do is figure out an extremely cool feature that breaks compatility with no workaround except a legacy data interpreter, so that the stalemate of legacy data interpretation can finally be broken.

**Comment by tehhowch** [comment]
Directly related to your example (and offered to illustrate that getting hung up on needing a legacy data interpreter isn't worthwhile at this time), but hardpoint firing angle restrictions can easily be added without any change to existing data files.

The current Hardpoint class constructor:
https://github.com/endless-sky/endless-sky/blob/e0d1780db0f6e13590d5ba7281106a5c197bf91b/source/Hardpoint.cpp#L43-L47
and the relevant calling code from Ship::Load (I omit the relevant Armament functions but you can look them up):
https://github.com/endless-sky/endless-sky/blob/e0d1780db0f6e13590d5ba7281106a5c197bf91b/source/Ship.cpp#L162-L191
That is, the currently supported hardpoint specification syntaxes are:
```
(gun | turret) X Y <outfit>
    ...
```
and
```
(gun | turret) <outfit>
    ...
```
Where `...` represents child nodes of the DataNode. In other words, we can simply use the child nodes of this `gun` / `turret` DataNode to handle extra information that needs to get passed to the constructor, such as `cone 30` or `priority 1` or `...`. Currently, nothing is done with these nodes (since there is no supporting AI logic that can use these *features*, because these *features* don't exist).
There's also nothing stopping us from adding `hardpoint` to the allowed keywords that works like `gun` and `turret` currently do (though any commit that does so will need to include how one specifies if the hardpoint represents a gun or a turret slot).

TLDR: Having a "legacy code loader" is not the hangup, because "legacy code" doesn't even necessarily exist. Implement the feature, and the best syntax for activating it can be found.

**Comment by accursedCursive** [comment]
When people use the term "spaghetti code", this is exactly what they mean.
Spaghetti code happens when new features are repeatedly added to something with minimal generalisation, with the result being a giant pile of nested conditionals necessary to handle every possibility.

If a hardpoint class is added but the armament function retains compatibility with legacy hardpoints, then that's going to add 6 more conditionals.
One to decide whether it's legacy or not so it can figure out if there's an extra argument for what kind of hardpoint it is.
Three more to handle the doubled number of possible syntaxes for the arguments (if there's an extra argument, it's the first argument, so all the other arguments have their indices shifted up).
Then two more to decide which kind of hardpoint it is when it comes time to add the hardpoint to the ship.

The result would be approximately twice as long as the current function for loading hardpoints.
I call the current function spaghetti code, but compared to the new function I just described, the current function is clean and simple.

If you drop compatibility with legacy hardpoints, then that extra complexity is avoided.
That lack of compatibility can then be compensated for with a legacy data interpreter.
Remember, the legacy data interpreter wouldn't just affect this case; it'd affect any case like this one, so in total it could save hundreds of lines of code and make important classes much more simple.

Another situation where a legacy data interpreter would simplify a function is if carriers were changed to allow any ship to be carried as long as the mass of that ship hull was within limits specified for the bay.
There'd be no need for the carrier function to consider the possibility of docking bays being specified as "fighter" or "drone" bays.
It'd get "drone x y" as something like:
dock 40 x y
and "fighter x y" as:
dock 60 x y

**Comment by tehhowch** [comment]
A few things:
 - You say "legacy hardpoint" but that isn't a thing - the game can use whatever internal representation it needs. What those who contribute to the game strive to do is maintain data file compatibility.
 - A Hardpoint class **already exists** (I linked you to it), so saying "If a hardpoint class is added" makes no sense. You can be more specific with your words to avoid these ambiguities and misunderstandings, so please do so.
 - Your "deconstruction" of my example indicates you don't have any experience writing or using the game's DataNode class, or an in-depth understanding of this game's code. All of this game's code is public, so before criticizing it I do suggest you take the time to understand it.
    - If the Hardpoint class gains new functionality, **of course** the Armament class functions AddGunPort and AddTurret are going to be updated to take a parameter that governs the new functionality. For example, if we added support for restricting the available angle of a hardpoint, AddGun and AddTurret would then take an `Angle` parameter that is defaulted to 360. (A turret would need an additional Angle to indicate its allowed neutral position, while it would be a matter of the actual feature being added if guns could use this parameter as well.) The hardpoint loading code could then look like this:
```
else if(key == "gun" || key == "turret") 
{ 
	if(!hasArmament) 
	{ 
		armament = Armament(); 
		hasArmament = true; 
	} 
	const Outfit *outfit = nullptr; 
	Point hardpoint; 
	if(child.Size() >= 3) 
	{ 
		hardpoint = Point(child.Value(1), child.Value(2)); 
		if(child.Size() >= 4) 
			outfit = GameData::Outfits().Get(child.Token(3)); 
	} 
	else if(child.Size() >= 2) 
		outfit = GameData::Outfits().Get(child.Token(1)); 
	Angle cone = Angle(360.);
	Angle aim;
	for(const DataNode &grand : child)
	{
		if(grand.Token(0) == "cone" && grand.Size() >= 2)
			cone = Angle(child.Value(1));
		else if(grand.Token(0) == "aiming angle" && grand.Size() >= 2)
			aim = Angle(child.Value(1));
	}
	if(outfit) 
		++equipped[outfit];
	if(key == "gun") 
		armament.AddGunPort(hardpoint, outfit, cone);
	else
		armament.AddTurret(hardpoint, outfit, aim, cone);
	// Print a warning for the first hardpoint after 32, i.e. only 1 warning per ship. 
	if(armament.Get().size() == 33) 
		child.PrintTrace("Warning: ship has more than 32 weapon hardpoints. Some weapons may not fire:"); 
}
```

Regarding your new example of letting any class be carried: I would implement this with only a single new key would be used (bay). The ship class would then be the next keyword on that line, e.g. `bay "Fighter"/fighter` or `bay "Drone"/drone` or `bay "Heavy Warship"`, with the remainder of that line being fully compatible with the existing code (after a conditional index shift).
The simplest implementation of a bay mass limit (which is a separate feature whose implementation is unrelated to that which allows any ship to be carried) would be to use a child node of the parent bay specification node, rather than try to add it to that already crowded data line.


 - You mention that the current subset of Ship::Load that I have posted here is "spaghetti code", so I invite you to provide the example C++ code that you feel should replace it, which has the generalisation required to provide support for the "cone" and "aiming angle" parameters.
 - You also haven't mentioned at all how your legacy data interpreter will interface in the game's startup process (GameData::BeginLoad() and GameData::LoadFile() so you can look up the current code). Are you proposing that the game first calls this "interpreter" to read every file, then writes the "proper" format, and then finally begins loading that? To me, that sounds like adding a feature requires even more code, because you'd have to update the interpreter to handle the current particular load syntax, update the class load method with the new load syntax, and then update the interpreter to output the new load syntax.

**Comment by endless-sky** [comment]
Consider some real examples of how the game has been expanded over time:
* Originally, it wasn't possible to specify which weapon goes in which hardpoint. The hardpoint tags were extended so that the outfit name could be given as an optional argument, but it's still also possible to not specify an outfit.
* Drone and fighter bays were originally all forward-facing and internal. Tags were added to allow "over" and "under" bays. Later the tags for facing direction were added. Again, backward compatibility was maintained.

In general it's possible to find a way to add new features without breaking backward compatibility, because the data parsing is designed to just skip anything it doesn't recognize.

**Comment by accursedCursive** [comment]
This is what ideal hardpoint loading would look like:

```
else if(key == "hardpoint") 
 { 
 	if(!hasArmament) { 
 		armament = Armament(); 
 		hasArmament = true; 
 	} 
	
 	string hardpointName;
 	double hardpointX = 0;
 	double hardpointY = 0;
	double hardpointRotation = 0;
 	string outfitName;
 	int outfitQuantity = 1;

	//This section is pseudocode due to my lack of C++ knowledge
	variablePointers = [hardpointName, hardpointX, hardpointY, rotation, outfitName, outfitQuantity];
	dataTypes = ["string", "number", "number" "number", "string", "number"];
	int currentArgument = 1;
	for (i = 0, dataTypes.Size()){
		if (type(child.Value(currentArgument)) == dataTypes[i])
			variablePointers[i] = currentArgument++;
	}

 	const Hardpoint *hardpoint = GameData::Hardpoints().Get(hardpointName); //This is intended to create a new instance of a template hardpoint that already comes with certain given properties, I don't know if this is the correct way to do this
 	const Outfit *outfit = nullptr; 
 	if(outfitName) 
 	 	outfit = GameData::Outfits().Get(outfitName); 
		++equipped[outfit]; 
		hardpoint->Outfit(outfit);

	hardpoint->OutfitQuantity(outfitQuantity); //Reload of a hardpoint is divided by this
	hardpoint->Position(Point(hardpointX, hardpointY));
	hardpoint->Rotation(hardpointRotation);

 	armament.AddHardpoint(hardpoint); 
} 
```

This is a much cleaner and more flexible way of loading hardpoints.
However, it cannot easily be made to support loading legacy hardpoints: This loading system expects that if the first argument is a string, then it's describing the type of hardpoint, but in rare cases a legacy hardpoint will have an outfit as the first argument.
These are the kinds of code cleanups you could do if you had a legacy data interpreter.

**Comment by tehhowch** [comment]
I'm of the opinion that that structure isn't necessary for this game. The DataNode class provides significant flexibility already - it can iterate through the tokens on a line, and the DataNodes on the indented lines below it - each of which can also be stepped through horizontally or vertically. You seem to only consider the horizontal dimension.

(On a side note, the type returned by `DataNode::Value(<index>)` is always `double`, and the type returned by `DataNode::Token(<index>)` is always `string`. There is a conditional check `DataNode::IsNumber(<index>)`, but your method would require we check that for every token, and also would disallow giving things numerical names (something not imposed by the current methods).)

Thank you for providing an example. I hope you are able to understand that there isn't anything currently in the game or proposable for the game that would benefit from overhauling every `::Load(<DataNode>)` method and creating a preprocessor to generate some arbitrary format.

I'll return to my original point: the thing which holds back features being added is feature implementations being proposed. This is a one-person show, and that one person has a full-time job and life outside of the hobby he has of writing code and developing both art and story for this game. If you or anyone really wants a feature, the fastest way to get it in game is to do it yourself, and propose it for review via Pull Request. The next fastest is to convince me or another contributor that your idea is worthwhile and useful, and hope we choose to spend our time on your idea.

**Comment by accursedCursive** [comment]
I don't see why you'd need to overhaul every DataNode method.
All you need is a preprocessor.
A preprocessor could potentially become a very powerful development tool, and save huge amounts of time in the long run.
And it doesn't even need to be something new: A more robust and flexible data loading system could act as a preprocessor.
A partially upgraded data loading system would still be better, so the implementation of a preprocessor doesn't have to be done all at once.

The first thing that would be needed is name wildcards.
Currently, you define/modify something by writing something like `ship "Model 32"` and then writing stuff inside that. With wildcards, you could write `ship "Model "* to modify every Kor Mereti ship, or `ship *` to modify every ship. Ideally it would allow wildcards and strings to come in any order, so you can write `outfit *"Plasma Thruster"` to modify every plasma thruster.

**Comment by tehhowch** [comment]
...**Why** would we ever need to modify game data like that to load information into the game?

You are describing a file format converter, I get that, but what I and MZ have said is there is no reason we need such a thing for the game.

Going back to your example, in which you create a stock **collection** of `Hardpoint` objects which can be loaded as root level nodes in a data file:
 - This would result in a `Hardpoint::Load(<DataNode>) method, which would be used anytime the game encounters the DataNode keyword "hardpoint" with a name token following (and optionally with 2 more tokens describing the x-y position, which would be given when referencing that hardpoint from a ship definition).
 - Encountering `gun` or `turret` in a ship definition would trigger a `Hardpoint(...)` default constructor call. In that constructor we would know that this is a specific gun or turret being made, and it would receive all the relevant parameters from the line and child nodes, as is done now.
 - the `Armament` class would likely be changed from using two add methods to just one `AddHardpoint(<Hardpoint>, <Outfit>)` method.
 - any future updates to hardpoint functionality syntax would be handled exclusively by `Hardpoint::Load`

So, I **really** don't see the purpose of writing a legacy data interpreter because 1) there is no legacy data syntax currently, and 2) we can extend the game ***without creating "legacy" data syntax***

**Comment by accursedCursive** [comment]
The number of arguments used when defining a hardpoint is one more than the number of arguments when defining a gun or turret. This makes it difficult to handle every syntax for the arguments while still retaining a small amount of code.
I guess you could define an index offset loading the arguments.

But hear me out on improving the way data loads to the point that it can be used as a legacy data interpreter: One thing you can't easily test in this game is balance changes, because there's so many ships and outfits that you have to change.

Say that you wanted to do weird balance experiments and have it so that all ships have 200 shields, but twice as much hull:
```
ship *
	shields 200
	hull a1*2 # a1 means the first argument of the tag being modified. If tags are applied and forgotten as soon as they are loaded, this won't work. * is a mathematical operator here.
```

Or have it so that all weapons have a default hardpoint sprite:
```
outfit *
	weapon
		"hardpoint sprite" a1 "hardpoint/default" # If a hardpoint sprite already exists, a1 will be the first argument. Otherwise, a1 is skipped and the default hardpoint sprite will be the first argument.
```

Replace all the asteroids with minables:
```
system *
	replace asteroids minables a1 a2 a3 # replace replaces a tag, like remove removes a tag
```

Or, of course, replace turrets and gun ports with hardpoints:
```
ship *
	replace gun hardpoint gun
	replace turret hardpoint turret

outfit *
	replace "gun ports" hardpoint gun
	replace "turret mounts" hardpoint turret
```

It's an extremely versatile system, and it wouldn't be difficult to make.
Except that I'm guessing tags are applied and forgotten as soon as they're loaded, which means they no longer have an argument list. If that's the case, the hardest part about this would be changing it so that tags are applied after loading.

**Comment by jafdy** [comment]
What's the relevance of your examples? Just create a simple program to do the replace. I am still slightly baffled as to what this will achieve, for one your comments seem to be targeted at replacing `gun` and ` turret` with a general `hardpoint` and for two you keep stating it is easy while failing to provide any example of a "legacy data interpreter" except in your one case.

**Comment by tehhowch** [comment]
Why do we need to build into Endless Sky (a game) a program that replaces or modifies the values read in from datafiles (a data writer)? Changing the **values** of data (which is fundamentally different than achieving the same game state when reading an alternate **syntax** of data specification) is easy to do with a simple program written in your language of choice. Furthermore, because of the current policy that we **do not** implement breaking changes (the first value after `"hull"` in a ship's `"attributes"` list is always going to be its maximum base hull value, or the first value after `"gun"` when the line has a length greater than 2 is always an X coordinate of that gun's position in the sprite pixel map), you don't have to worry about changes in this repo making you rewrite your data modifier program.

In fact, a fellow contributor has already written an Endless Sky datafile generator in Python. It should be trivial to extend his offering to perform rapid alterations to things which you deign to alter, and write out properly formatted datafiles which you can use to simply and easily test balance changes.
https://github.com/comnom/ES-tools

I recommend closing this issue:
1. There's no data in this repo that is specified incorrectly
2. There's no plan to create data which is specified incorrectly
3. There's no plan to implement features that make data be specified incorrectly
4. Your goal is to modify datafile values
    -  Use a text editor to modify the values, or write a program / script that does find + replace with your desired new values.

**Comment by accursedCursive** [comment]
You're being rather short-sighted with this: You can't see the usefulness of these features, because no one is taking advantage of these features (because, of course, they don't exist).

With this small change, it would be possible for anyone to effortlessly change how the game is balanced.
This would have a serious impact on how the game is developed: Any developer could easily try out a dozen majorly-different versions of Endless Sky in an hour, and share their parameters with other developers.
Eventually, developers will agree that a certain set of parameters is much better than the vanilla game, and those parameters will be properly applied.

Plus of course, you can abandon the breaking change policy since breaking changes would become pretty much impossible.

**Comment by Hadron1776** [comment]
>With this small change, it would be possible for anyone to effortlessly change how the game is balanced.
This would have a serious impact on how the game is developed: Any developer could easily try out a dozen majorly-different versions of Endless Sky in an hour, and share their parameters with other developers.
Eventually, developers will agree that a certain set of parameters is much better than the vanilla game, and those parameters will be properly applied.

Yup. Just like those ~~developers~~ contributors could easily make this small change on their own forks, and once they agree on an implementation, they can make a PR for it.

(If you want this done so badly, you really should try making a PR for it)

**Comment by accursedCursive** [comment]
This is a text parsing and data management job.
I suck at text parsing, and haven't worked with Endless Sky's data management other than adding a new tag to my fork.

Maybe I'll do it myself at a _much_ later date, but I would appreciate it if someone else did this so I can focus on less finnicky stuff while I get the hang of C++.

**Comment by tehhowch** [comment]
>This is a text parsing and data management job

Indeed. It is not a job for the game. Note that this repository consists of only the content used by the game. It does not include tools or utilities that are used to create data used by the game. There are other repositories for that sort of thing.

**Comment by accursedCursive** [comment]
In case you haven't realised, this is a request for very streamlined automatic data modification.
A third-party Python tool which would have to be reprogrammed to be useful for this is hardly streamlined.
Having this adaptability written into the game itself is very streamlined.
Plus, it's absolutely a job for the game: It's a huge moddability improvement for a game reliant on modders.

Also, it's not a tool or utility. It's merely a few more syntax rules for game data, that so happen to allow you to make huge changes to the game just by adding a few lines of data.
Temporary changes of course, it's experimental and if a change turns out to be good it is manually applied.

It has an extremely high reward:difficulty ratio as far as features go, it merely requires someone with a good understanding of C++ and Endless Sky to make these changes:
-Allow multiple definitions to be modified at once through the use of wildcards
-Buffer data before loading it (Unless this is already done) to make the next change possible
-Allow the original arguments of a tag being modified to be referred to when modifying the tag, and allow mathematical operators to be used on them
-Introduce a "replace" tag so a tag can be replaced with another tag as it is modified, and make the "remove" tag work for everything instead of just systems or whatever
I only know enough to do about half of that.

It'd be loads of fun having a feature like this, people could come up with all kinds of strange modifications to the game.

**Comment by jafdy** [comment]
Given a day I could create a tool that would allow you to do this just fine, I could even give it a graphical interface. I am struggling to understand why this is needed.
In general moders are able to handle a bit of grunt work, it comes with content creating, and with tools already available to use there is no need to make a change to the game.

**Comment by Wrzlprnft** [comment]
i recommend splitting this in the future in the actual issues:
- a pre loading file format system, so one can do wonky things
- a name wildcard
so they can be discussed seperatly. it seems like the legacy data loader could be a way to implement a name wildcard, but is not necessarily the way to do it, and a name wildcard is (hopefully) not the only example of how data formatting would benefit the game. the way it is, it is hard to discuss and solve/close them seperatly.

i think you may also benefit from some total conversion support, which has, as far as i know, not happened yet. basically a keyword to ignore the vanilla files. i think, of you'd change the values yourself via e.g. a quick script, and put those in the plugin folder, it would achieve a similar effect and prevent vanilla game changes from messing your balance up.

Don't forget, you have to convince people your ideas pay off, even more if you can't do it yourself. i'd change from a “you need“ to a “maybe we could“

**Comment by accursedCursive** [comment]
@jafdy The goal is to maximise the ease and efficiency of bulk modding.
The highest achievable ease and efficiency comes from data loading with extended functionality.
The reason why this is a desirable goal is because when absolutely everyone can heavily mod the game just by sticking a simple text folder in the plugins, there will be a period where a good portion of the ES community is experimenting with totally different game setups.
Eventually, people will find setups that solve the balance issues with Endless Sky, and also make Endless Sky much more fun.

While this may sound like an unnecessarily difficult solution to a problem that can easily be solved without programming and instead just grunt work, making it extremely easy for everyone to figure out for themselves how something should be changed is much easier than convincing them of how something should be changed.
It'll save work in the long run, and guarantee that ES _eventually_ gets its much-needed combat balance overhaul.

**Comment by comnom** [comment]
> The goal is to maximise the ease and efficiency of bulk modding.

Creating a special syntax to bulk modify the data files is of limited usefulness. They're just text files, any simple scripting language can modify them with ease. In my planet sprites PR, I modify over a thousand lines with a script that took a whopping five minutes to write, using the python data parser/writer tehhowch mentioned. (This is easily equal to or less than the time it would take to use your hypothetical new syntax to do it.)


As for testing large scale balance changes, that's completely counter-intuitive to balance testing. You're not going to get any useful metrics out of a single person testing large scale changes in under an hour, end of story.

**Comment by accursedCursive** [comment]
Equal to or less than?
With my hypothetical syntax you could do it in a matter of seconds, with no programming knowledge.
To use an earlier example, if you want to replace all asteroids with minables (assuming you've added new minables of the appropriate names):
```
system *
	replace asteroids minables a1 a2 a3
```
Clearly, you didn't read past the line you quoted, since you ignored my reasoning for why this should be a feature as opposed to any workaround.
Besides, workarounds like yours must take almost as much effort as improving the internal data parsing, assuming they provide the same range of functionality as my suggested improvements.

**Comment by comnom** [comment]
> With my hypothetical syntax you could do it in a matter of seconds, with no programming knowledge.

Five minutes was to match specific sprites to their variant, and replace half using a random sample.

You'd have to make a more convincing argument as to why I can't just use a trivial find and replace script to do what you're suggesting, and even then, despite tehhowch's best efforts, you refuse to acknowledge that that kind of text editing is outside the scope of the game. (ie. a video game does not need to edit text files beyond writing a save.)

> Besides, workarounds like yours...

It's an ancilliary tool that provides functionality that doesn't belong in a video game.

>  must take almost as much effort as improving the internal data parsing, assuming they provide the same range of functionality as my suggested improvements.

I see no improvements to the the data parser in any of your posts. You started with a convoluted explanation of working around the parser with an interpreter, and now you want scripting syntax that can be found in most shell environments.

**Comment by accursedCursive** [comment]
Sounds like you're opposing this for the sake of participating in the community.
You've skimmed through the thread, completely missed the purpose and specific details of the suggestion, and now expect to have everything explained to you.
This behaviour isn't helping anyone, everyone else in this thread had the decency to read and understand before talking.
And before you skim through it a second time, I should point out that I've never seen anyone switch from making completely off-topic arguments to suddenly understanding well enough to give insightful input.

**Comment by Wrzlprnft** [comment]
One problem which i see with data manipulation in a script like manner rather than just data loading is on how to handle priorities. as e.g. used in rimworld and endless sky, load order is sufficient to determine priority. what with scripts, do them step by step? just accept only one such modification of the files?
i can see that it allows for a somewhat quicker implementation on the side of the modder, 

but since those changes are all relative, a lot more potential f#ckups, and a lot of edge cases to discuss.

sidenote:
with the introduction of a ubuntu shell on windows and the ability to really easily use grep and co to make those permanent data file changes, i do not see a real benefit of learning your data interpreters syntax over a hand ful of command like tools which are well documented.

**Comment by accursedCursive** [comment]
The changes would be applied in the current order data is loaded.
So, if you wanted to apply a change to systems, you do one of three things:

- Put your modification in a plugin
This option is what most users should do.

- Put your modification at the end of systems.txt
This option makes the most sense for developers making a change to systems in the vanilla game, but not hardcoding it because they aren't sure about it.

-Put your modification in a file called zzzzzz.txt 
This option makes the most sense for having a built-in set of parameters for Endless Sky (which doesn't make much sense).

Also, this system is much more intuitive and simple than command line tools, because it's basically just making a plugin but with a few more keywords so you can do more with less effort.

**Comment by jafdy** [comment]
A modder that is allergic to a small amount of work probably doesn't have the dedication to make a mod big enough to warrant this kind of support. And once again, you have failed to explain why we shouldn't just use a tool, or at least not in any way I find comprehensible.

**Comment by jafdy** [comment]
Why will this encourage experimentation? Just providing a different way to do it won't encourage anything unless the will was already there, and it clearly isn't there because the current approaches are not hard. This is especially true is you limit yourself to smaller groups of ships initially.

**Comment by MCOfficer** [comment]
i can very well imagine that being able to test with just 3 lines instead of 100 will increase the amount of people playing around with mechanics. at the moment, at least for new modders, modifying a large chunks of the game's data is rightfully considered a project, and not just an hour's work.
